packages/strapi-generate-api/templates/mongoose/service.template

```diff
@@ -9,6 +10,144 @@
 // Public dependencies.
 const _ = require('lodash');
 
+const buildTempFieldPath = field => {
+  return `__${field}`;
+};
+
+const restoreRealFieldPath = (field, prefix) => {
+  return `${prefix}${field}`;
+};
+
+export const generateLookupStage = (
+  strapiModel,
+  { whitelistedPopulate = null, prefixPath = '' } = {}
+) => {
+  const result = strapiModel.associations
+    .filter(ast => {
+      if (whitelistedPopulate) {
+        return _.includes(whitelistedPopulate, ast.alias);
+      }
+      return ast.autoPopulate;
+    })
+    .reduce((acc, ast) => {
+      const model = ast.plugin
+        ? strapi.plugins[ast.plugin].models[ast.collection || ast.model]
+        : strapi.models[ast.collection || ast.model];
+
+      const from = model.collectionName;
+      const isDominantAssociation = ast.dominant || !!ast.model;
+
+      const _localField = !isDominantAssociation
+        ? '_id'
+        : ast.via === strapiModel.collectionName || ast.via === 'related'
+          ? '_id'
+          : ast.alias;
+
+      const localField = `${prefixPath}${_localField}`;
+
+      const foreignField = ast.filter
+        ? `${ast.via}.ref`
+        : isDominantAssociation
+          ? ast.via === strapiModel.collectionName
+            ? ast.via
+            : '_id'
+          : ast.via === strapiModel.collectionName
+            ? '_id'
+            : ast.via;
+
+      // Add the juncture like the `.populate()` function
+      const asTempPath = buildTempFieldPath(ast.alias, prefixPath);
+      const asRealPath = restoreRealFieldPath(ast.alias, prefixPath);
+      acc.push({
+        $lookup: {
+          from,
+          localField,
+          foreignField,
+          as: asTempPath,
+        },
+      });
+
+      // Unwind the relation's result if only one is expected
+      if (ast.type === 'model') {
+        acc.push({
+          $unwind: {
+            path: `$${asTempPath}`,
+            preserveNullAndEmptyArrays: true,
+          },
+        });
+      }
+
+      // Preserve relation field if it is empty
+      acc.push({
+        $addFields: {
+          [asRealPath]: {
+            $ifNull: [`$${asTempPath}`, null],
+          },
+        },
+      });
+
+      // Remove temp field
+      acc.push({
+        $project: {
+          [asTempPath]: 0,
+        },
+      });
+
+      return acc;
+    }, []);
+
+  return result;
+};
+
+export const generateMatchStage = (
```

Comment:
// TODO: Rewrite this logic to be easier to read and maintain afterward.
The problem with the current logic is that the first relation doesn't get populated !!